services/console-proxy/server/src/main/java/com/cloud/consoleproxy/ConsoleProxyNoVncClient.java

```diff
@@ -137,18 +145,80 @@ public void run() {
 
     /**
      * Authenticate to VNC server when not using websockets
+     *
+     * Since we are supporting the 3.8 version of the RFB protocol, there are changes on the stages:
+     * 1. Handshake:
+     *    1.a. Protocol version
+     *    1.b. Security types
+     * 2. Security types
+     * 3. Initialisation
+     *
+     * Reference: https://github.com/rfbproto/rfbproto/blob/master/rfbproto.rst#7protocol-messages
      * @throws IOException
      */
     private void authenticateToVNCServer() throws IOException {
-        if (!client.isVncOverWebSocketConnection()) {
+        if (client.isVncOverWebSocketConnection()) {
+            return;
+        }
+
+        if (client.isVncOverTunnel()) {
             String ver = client.handshake();
             session.getRemote().sendBytes(ByteBuffer.wrap(ver.getBytes(), 0, ver.length()));
 
-            byte[] b = client.authenticate(getClientHostPassword());
+            byte[] b = client.authenticateTunnel(getClientHostPassword());
             session.getRemote().sendBytes(ByteBuffer.wrap(b, 0, 4));
+        } else {
+            ByteBuffer verStr = client.handshakeProtocolVersion();
+            sendMessageToVNCClient(verStr.array(), 12);
+
+            int secType = client.handshakeSecurityType();
+            byte[] numberTypesToClient = new byte[] { 1, (byte) secType };
+            sendMessageToVNCClient(numberTypesToClient, 2);
+
+            client.processHandshakeSecurityType(secType, getClientHostPassword(),
+                    getClientHostAddress(), getClientHostPort());
+
+            byte[] securityResultToClient = new byte[] { 0, 0, 0, 0 };
+            sendMessageToVNCClient(securityResultToClient, 4);
+            client.setWaitForNoVnc(true);
+
+            while (client.isWaitForNoVnc()) {
+                s_logger.debug("Waiting");
+            }
```

Comment:
do we want this repetitive single word message in the log file? It seems that it would not be helpful during debugging by an operator. I would sugest one start message and than keeping track of time for a second message:
```suggestion
            s_logger.debug("Waiting for NoVnc");
            int cycles = 0;
            while (client.isWaitForNoVnc()) {
                cycles++
            }
            if (s_logger.isDebugEnabled()) {
                s_logger.debug(String.format("Waited %d cycles for NoVnc", cycles));
            }
```

Reply:
Done, thanks